### PR TITLE
Unify tokenizers

### PR DESCRIPTION
**Pull request by andrewPoulton**
Attempts to address #308, altering tokenizer path to a single universal hf-format file.

Since a bunch of user code would inevitably have the old merges/vocab filepaths baked in, I included a script to unify the old format to the new, and bring up an error (with the mitigation steps) if the new format is not used.

Note - this does not address hardcoded args for the API, as the task is set to `language_modelling` there.

### CODE REVIEW

File: tests/test_streaming_language_modeling_task.py [hunk 1]
```diff
@@ -41,18 +44,21 @@ def get_last_log_stats_containing_string(log_records, search_string):
...
             with self.assertLogs() as logs:
                 train_language_model(
                     data_dir=data_dir,
                     arch="transformer_lm_gpt2_tiny",
                     extra_flags=[
-                        "--vocab-filename",
-                        vocab,
-                        "--merges-filename",
-                        merges,
+                        # "--vocab-filename",
```
Comment (suchenzang) [inline tests/test_streaming_language_modeling_task.py:56]: Nit: cleanup commented out stuff?

File: tests/test_streaming_language_modeling_task.py [hunk 2]
```diff
@@ -68,7 +74,7 @@ def get_last_log_stats_containing_string(log_records, search_string):
                     task="streaming_language_modeling",
                     max_tokens=None,
                 )
-            train_log = get_last_log_stats_containing_string(logs.records, "train_loss")
+            train_log = get_last_log_stats_containing_string(logs.records, '"loss"')
```
Comment (suchenzang) [inline tests/test_streaming_language_modeling_task.py:77]: Was this test not running before? How did these log lines all get renamed...

File: metaseq/optim/lr_scheduler/fixed_schedule.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+from dataclasses import dataclass, field
```
[the remark this answers is not in the document]
Comment (andrewPoulton) [reply metaseq/optim/lr_scheduler/fixed_schedule.py:6]: It snuck in at https://github.com/facebookresearch/metaseq/pull/429/commits/d9d3b1ba1c59e1bd0955399e8de9bd399d4b2736, it's needed for tests/test_streaming_language_modeling.py (which isn't run as part of any CI/GH action workflow, so might be deprecated?). Happy to remove if preferred?

File: tests/test_streaming_language_modeling_task.py (hunk 2 above)
Comment (andrewPoulton) [reply tests/test_streaming_language_modeling_task.py:77]: Yeah, it doesn't run as part of CI/actions builds (the test_streaming_language_modeling_task.py under cpu_tests/ is run, but not this one, and they're different). The logs produced here don't contain 'train_loss' (anymore).

File: metaseq/optim/lr_scheduler/fixed_schedule.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (suchenzang) [reply metaseq/optim/lr_scheduler/fixed_schedule.py:6]: Yeah let's get rid of this file - there's a few hacks sprinkled in throughout the codebase to root out references to this "fixed" LR scheduler; haven't been able to cleanly remove these references just yet... opening https://github.com/facebookresearch/metaseq/issues/434 to track this

File: metaseq/models/transformer_decoder.py [hunk 4]
```diff
@@ -119,7 +119,8 @@ def __init__(self, args, dictionary, embed_tokens):
             if args.decoder_learned_pos and not self.use_alibi
             else None
         )
-        self.embed_positions.to(device).to(dtype)
+        if self.embed_positions is not None:
```
Comment (suchenzang) [inline metaseq/models/transformer_decoder.py:122]: We had cases where `self.embed_positions` were None at this point?

File: metaseq/optim/base_optimizer.py [hunk 5]
```diff
@@ -108,7 +108,9 @@ def multiply_grads(self, c):
                     c = c.to(p.grad.device)
                 p.grad.data.mul_(c)
 
-    def clip_grad_norm(self, max_norm, norm_type="l2", aggregate_norm_fn=None):
+    def clip_grad_norm(
+        self, max_norm, norm_type="l2", aggregate_norm_fn=None, **kwargs
```
Comment (suchenzang) [inline metaseq/optim/base_optimizer.py:112]: Oh no - where did these kwargs come from?

File: metaseq/models/transformer_decoder.py (hunk 4 above)
Comment (andrewPoulton) [reply metaseq/models/transformer_decoder.py:122]: Yeah -L119 is the condition for no embed_positions - happens for `transformer_lm_gpt2_tiny` at least. But the `decoder_learned_pos` option defaults to False, so presumably this is usually overridden by user args?

File: metaseq/tasks/streaming_language_modeling.py [hunk 6]
```diff
@@ -122,17 +117,19 @@ class StreamingLanguageModelingTask(LegacyTask):
     """
 
     def __init__(self, args):
+        if not hasattr(args, "hf_tokenizer"):
+            msg = f"""`StreamingLanguageModelingTask` expects a unified tokenizer format, and not
+            the merges/vocab dual format from days of yore. You can convert the old format to the new by running
+            `python -m metaseq.scripts.unify_tokenizer <PATH_TO_TOKENIZERS>`
```
[the remark this answers is not in the document]
Comment (andrewPoulton) [reply metaseq/tasks/streaming_language_modeling.py:123]: Yeah, I thought of doing this, but backtracked because I didn't have write permission for one of the tokenizer locations on azure - can we assume that users always will?

[the remark this answers is not in the document]
Comment (andrewPoulton) [reply metaseq/tasks/streaming_language_modeling.py:123]: Maybe can try and catch the permissions error with a similar message?

File: metaseq/models/transformer_decoder.py (hunk 4 above)
Comment (suchenzang) [reply metaseq/models/transformer_decoder.py:122]: Oof... yeah... let's do the not-so-great thing and change that default to `True`.  We clobber this with our sweep scripts all the time so might as well bite the bullet here and make it consistent throughout (opened https://github.com/facebookresearch/metaseq/issues/435 to track; good work flagging all of these papercuts 😅  )

File: metaseq/tasks/streaming_language_modeling.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (suchenzang) [reply metaseq/tasks/streaming_language_modeling.py:123]: Oh we need to fix your permissions on Azure then - yeah we should assume users have write permissions to their data directories, but this needs to fail noisily (on job launch / slurm submission, before job is queued up and waiting for machines).

File: metaseq/optim/base_optimizer.py (hunk 5 above)
Comment (andrewPoulton) [reply metaseq/optim/base_optimizer.py:112]: Lol I don't know now - I remember some test failed but now I can't reproduce 🤷

Comment (andrewPoulton) [reply metaseq/optim/base_optimizer.py:112]: Oh wait I found it - again in the unused tests, the `MetaseqAdam` class inherits `clip_grad_norm` from the `BaseOptimizer`, which doesn't accept a `skip_gradient_update_on_clip_norm` argument, but this is passed in from the `Trainer` object.  I'll change the base optimizer to accept skip_gradient_update_on_clip_norm

Comment (suchenzang) [reply metaseq/optim/base_optimizer.py:112]: We should probably promote some of these unused tests to being used again, lol. Opening https://github.com/facebookresearch/metaseq/issues/436 to track

File: metaseq/tasks/streaming_language_modeling.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (andrewPoulton) [reply metaseq/tasks/streaming_language_modeling.py:123]: Actually I think there'll be a noisy failure anyway now if the user tries to pass any of the old args (like `--vocab-filename`), the argparser will flip out as it's not expecting that arg. I don't _think_ there's a tidy way to intercept it 🤷

[the remark this answers is not in the document]
Comment (suchenzang) [reply metaseq/tasks/streaming_language_modeling.py:123]: It'll probably be adding some logic here: https://github.com/facebookresearch/metaseq/blob/37809ec65632c434ed722a5c8aa37cd194ce9f8b/metaseq/launcher/slurm.py#L412

File: metaseq/launcher/opt_baselines.py [hunk 7]
```diff
@@ -81,6 +81,9 @@ def get_grid(args):
         if args.circleci:
             data_loc_by_env = "./gpu_tests/circleci"
             valid_subsets = ["BookCorpusFair"]
+            import metaseq.scripts.unify_tokenizer as unify_tokenizer
```
Comment (stephenroller) [inline metaseq/launcher/opt_baselines.py:84]: move to top?

File: metaseq/models/transformer_lm.py [hunk 8]
```diff
@@ -209,7 +209,7 @@ def base_lm_architecture(args):
     args.decoder_ffn_embed_dim = getattr(args, "decoder_ffn_embed_dim", 2048)
     args.decoder_layers = getattr(args, "decoder_layers", 6)
     args.decoder_attention_heads = getattr(args, "decoder_attention_heads", 8)
-    args.decoder_learned_pos = getattr(args, "decoder_learned_pos", False)
+    args.decoder_learned_pos = getattr(args, "decoder_learned_pos", True)
```
Comment (stephenroller) [inline metaseq/models/transformer_lm.py:212]: huh, why are we changing this?

File: metaseq/optim/lr_scheduler/__init__.py [hunk 9]
```diff
@@ -18,7 +18,7 @@
     LR_SCHEDULER_REGISTRY,
     LR_SCHEDULER_DATACLASS_REGISTRY,
 ) = registry.setup_registry(
-    "--lr-scheduler", base_class=BaseLRScheduler, default="fixed"
+    "--lr-scheduler", base_class=BaseLRScheduler, default="inverse_sqrt"
```
Comment (stephenroller) [inline metaseq/optim/lr_scheduler/__init__.py:21]: This seems unrelated and controversial

**Review by stephenroller — COMMENTED** [review]
I think we should revisit this?

File: metaseq/optim/base_optimizer.py [hunk 10]
```diff
@@ -108,8 +108,16 @@ def multiply_grads(self, c):
...
                 p.grad.data.mul_(c)
 
-    def clip_grad_norm(self, max_norm, norm_type="l2", aggregate_norm_fn=None):
-        """Clips gradient norm."""
+    def clip_grad_norm(
+        self,
+        max_norm,
+        norm_type="l2",
+        aggregate_norm_fn=None,
+        skip_gradient_update_on_clip_norm=False,
```
[the remark this answers is not in the document]
Comment (andrewPoulton) [reply metaseq/optim/base_optimizer.py:116]: Oh yeah, a lot of the unrelated stuff was to get tests working - this whole PR probably needs binning and restarting from the new head(s). Will take a look tomorrow.
